static/js/popover_menus_data.js

```diff
@@ -87,16 +87,11 @@ export function get_actions_popover_content_context(message_id) {
     const should_display_read_receipts_option =
         page_params.realm_enable_read_receipts && not_spectator;
 
-    function is_add_reaction_icon_visible() {
-        const $message_row = message_lists.current.get_row(message_id);
-        return $message_row.find(".message_controls.reaction_button").is(":visible");
-    }
-
     // Since we only display msg actions and star icons on windows smaller than
     // `media_breakpoints.sm_min`, we need to include the reaction button in the
     // popover if it is not displayed.
     const should_display_add_reaction_option =
-        !message.is_me_message && !is_add_reaction_icon_visible();
+        !message.is_me_message && Boolean(message.clean_reactions);
```

Comment:
So, this is a bit of an abstraction violation -- the `message.clean_reactions` field is an internal property for the `reactions.js` module. I think you could address this by extracting a `reactions.message_has_reactions(message)` function which returns `Boolean(message.clean_reactions.length === 0)` after doing a `set_clean_reactions` call to make sure its state is up-to-date.

I'm also not entirely sure the check is correct -- what does `clean_reactions` look like for a message which previously had reactions but they've all been removed?

Reply:
@timabbott, I have made changes for the first suggestion and created a separate function for it inside reactions.js. However, regarding the second question, the `clean_reactions` is a map that contains the type of reactions present in a message, not the number of reactions. Due to this, if a message previously had one reaction, the `clean_reactions` map would have a size of 1 at that moment, but when the reaction is removed, the map updates simultaneously to show a size of 0, which is the expected behavior.